modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/validation/Validator.java

```diff
@@ -0,0 +1,160 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.configuration.processor.validation;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessor.TOP_LEVEL_ANNOTATIONS;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.containsAnyAnnotation;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.isValidValueAnnotationFieldType;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.joinSimpleName;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.sameType;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.simpleName;
+
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.VariableElement;
+import org.apache.ignite.internal.configuration.processor.ClassWrapper;
+import org.apache.ignite.internal.configuration.processor.ConfigurationValidationException;
+
+/**
+ * Base class for configuration validators.
+ */
+public abstract class Validator {
+    final ProcessingEnvironment processingEnvironment;
+
+    Validator(ProcessingEnvironment processingEnvironment) {
+        this.processingEnvironment = processingEnvironment;
+    }
+
+    public abstract void validate(ClassWrapper classWrapper) throws ConfigurationValidationException;
+
+    @SafeVarargs
+    static void assertNotContainsFieldAnnotatedWith(ClassWrapper classWrapper, Class<? extends Annotation>... annotations) {
+        if (containsAnyAnnotation(classWrapper.fields(), annotations)) {
+            throw new ConfigurationValidationException(classWrapper, String.format(
+                    "Must not have fields annotated with %s.", joinSimpleName(" or ", annotations)
+            ));
+        }
+    }
+
+    @SafeVarargs
+    static void assertSuperClassNotContainsFieldAnnotatedWith(ClassWrapper classWrapper, Class<? extends Annotation>... annotations) {
+        if (containsAnyAnnotation(classWrapper.requiredSuperClass().fields(), annotations)) {
+            throw new ConfigurationValidationException(classWrapper, String.format(
+                    "Superclass %s must not have fields annotated with %s.",
+                    classWrapper.requiredSuperClass(), joinSimpleName(" or ", annotations)
+            ));
+        }
+    }
+
+    static void assertHasNoSuperClass(ClassWrapper classWrapper) {
```

Comment:
As far as I remember, we have a similar test in `ClassWrapper`, is it needed here?

Reply:
In `ClassWrapper` we have a `getRequiredSuperclass` method, which does the opposite, do you think we should move it there?

Reply:
Why not.

Reply:
Because this class contains all the `static` methods related to validation. I agree that having a validation method in `ClassWrapper` is not perfect, but I don't want to make it even more ugly

Reply:
It looks a bit strange, on the one hand in `ClassWrapper` there is a method with validation, you don't want to add a new one and we get 2 classes with validation, which can be confusing in the future. In my opinion, it is better to separate logically. I think it would be good if both methods were static and were in one Validator.

Reply:
Very nice!